include/aws/common/byte_buf.h

```diff
@@ -61,6 +62,65 @@ static inline struct aws_byte_buf aws_byte_buf_from_array(const uint8_t *c_str,
     return buf;
 }
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * No copies, no string allocations. Fills in output with a list of aws_byte_cursor instances where buffer is
+ * an offset into the input_str and len is the length of that string in the original buffer.
+ *
+ * Edge case rules are as follows:
+ * if the string begins with split_on, an empty string will be the first entry in output
+ * if the input string has two adjacent split_on tokens, an empty string will be inserted into the output.
+ * if the input string ends with split_on, an empty string will be appended to the output.
+ *
+ * It is the user's responsibility to properly initialize output. Recommended number of preallocated elements from output
+ * is your most likely guess for the upper bound of the number of elements resulting from the split.
+ *
+ * The type that will be stored in output is struct aws_byte_cursor (you'll need this for the item size param).
+ *
+ * It is the user's responsibility to make sure the input buffer stays in memory long enough to use the results.
+ */
+AWS_COMMON_API int aws_string_split_on_char(struct aws_byte_buf *input_str, char split_on,
+                                            struct aws_array_list *output);
+
+/**
+* No copies, no string allocations. Fills in output with a list of aws_byte_cursor instances where buffer is
+* an offset into the input_str and len is the length of that string in the original buffer. N is the max number of splits,
+* if this value is zero, it will add all splits to the output.
+*
+* Edge case rules are as follows:
+* if the string begins with split_on, an empty string will be the first entry in output
+* if the input string has two adjacent split_on tokens, an empty string will be inserted into the output.
+* if the input string ends with split_on, an empty string will be appended to the output.
+*
+* It is the user's responsibility to properly initialize output. Recommended number of preallocated elements from output
+* is your most likely guess for the upper bound of the number of elements resulting from the split.
+*
+* The type that will be stored in output is struct aws_byte_cursor (you'll need this for the item size param).
+*
+* It is the user's responsibility to make sure the input buffer stays in memory long enough to use the results.
+*/
+AWS_COMMON_API int aws_string_split_on_char_n(struct aws_byte_buf *input_str, char split_on,
+    struct aws_array_list *output, size_t n);
+
+/**
+ * Copies from to to. If to is too small, AWS_ERROR_DEST_COPY_TOO_SMALL will be returned.
```

Comment:
If the destination is too small, what is the state of the destination buffer after the call?

Reply:
It isn’t touched as the bounds check happens first.